### PR TITLE
Add ebcli Library; Override Version Label

### PR DESCRIPTION
**Pull request by whatisdot**
See #348 for motivations.

**For the sake of review, I have temporarily uploaded the ebcli to a private repository [located here](https://github.com/bodyshopbidsdotcom/ebcli_source).**

(also note that all CodeClimate issues are ebcli related, which I don't plan on refactoring)

This provides parallel deployment (contains commits from #348) with correct version labels by overriding the methods the eb command uses to generate them.

## How

I copied two methods from the original source (only available [through pypi](https://pypi.org/project/awsebcli/#files)), and modified the lines that pass the values used for version label and description. I also had to copy the module that loads the deploy handler, so that I could substitute our custom deploy handler.

Everything in `ebcli_overrides` is duplicate logic, due to the size of the methods and code structure. The changes are limited to the following:

### DeployController
Extend the original, override the [`do_command` method](https://github.com/bodyshopbidsdotcom/ebcli_source/blob/master/controllers/deploy.py#L52) so we save the message argument, and then call the original `do_command` method:
```python
class DeployController(OriginalDeployController):
  def do_command(self):
    self.message = self.app.pargs.message
    super(DeployController, self).do_command()

# ...within the `compose_deploy` method...
# define the label, then pass the previously omitted label and message to the `create_app_version` method
      label = module + '-' + SourceControl.get_source_control().get_version_label()
      version_label = commonops.create_app_version(app_name, label=label, message=self.message, process=process_app_version)
```

### ebcore
This is copied from the [original ebcore file](https://github.com/bodyshopbidsdotcom/ebcli_source/blob/master/core/ebcore.py).
```python
from ..controllers.deploy import DeployController
# from ebcli.controllers.deploy import DeployController
```

## Results

Description will remain the same: the tag associated with the docker image. The application version will now have a prefix of the environment name to clarify which configuration was deployed to which environment.

screen shot 2018-08-02 at 2 59 46 pm

Deploys from the `git_push_eb` command will deploy to all environments simultaneously.

```bash
$ git_push_eb stage
# ...output from original build, followed by new output:

--- Configuring module: dc30-qa2-web ---

--- Configuring module: dc30-qa2-global-worker ---

--- Configuring module: dc30-qa2-callbacks-worker ---
[2018-08-02 14:49:52,922]: Deploying to all web and worker envs.
--- Creating application version for module: dc30-qa2-web ---
[2018-08-02 14:49:53,355]: Found credentials in shared credentials file: ~/.aws/credentials
[2018-08-02 14:49:53,412]: Starting new HTTPS connection (1): elasticbeanstalk.us-west-2.amazonaws.com
[2018-08-02 14:49:53,798]: Starting new HTTPS connection (1): elasticbeanstalk-us-west-2-712793492101.s3.us-west-2.amazonaws.com
Uploading snapsheet-url-shortener/dc30-qa2-web-app-180802_144953.zip to S3. This may take a while.
Upload Complete.
--- Creating application version for module: dc30-qa2-global-worker ---
[2018-08-02 14:49:54,902]: Found credentials in shared credentials file: ~/.aws/credentials
[2018-08-02 14:49:54,960]: Starting new HTTPS connection (1): elasticbeanstalk.us-west-2.amazonaws.com
[2018-08-02 14:49:55,443]: Starting new HTTPS connection (1): elasticbeanstalk-us-west-2-712793492101.s3.us-west-2.amazonaws.com
Uploading snapsheet-url-shortener/dc30-qa2-global-worker-app-180802_144954.zip to S3. This may take a while.
Upload Complete.
--- Creating application version for module: dc30-qa2-callbacks-worker ---
[2018-08-02 14:49:56,406]: Found credentials in shared credentials file: ~/.aws/credentials
[2018-08-02 14:49:56,457]: Starting new HTTPS connection (1): elasticbeanstalk.us-west-2.amazonaws.com
[2018-08-02 14:49:56,860]: Starting new HTTPS connection (1): elasticbeanstalk-us-west-2-712793492101.s3.us-west-2.amazonaws.com
Uploading snapsheet-url-shortener/dc30-qa2-callbacks-worker-app-180802_144956.zip to S3. This may take a while.
Upload Complete.
--- Waiting for Application Versions to be pre-processed ---
Finished processing application version dc30-qa2-global-worker-app-180802_144954
Finished processing application version dc30-qa2-web-app-180802_144953
Finished processing application version dc30-qa2-callbacks-worker-app-180802_144956
--- Creating modules ---
               dc30-qa2-callbacks-worker - INFO: Environment update is starting.
               dc30-qa2-callbacks-worker - INFO: Deploying new version to instance(s).
                            dc30-qa2-web - INFO: Environment update is starting.
               dc30-qa2-callbacks-worker - INFO: Starting new ECS task with awseb-dc30-qa2-callbacks-worker-ze3wbf4gfx:14.
               dc30-qa2-callbacks-worker - INFO: ECS task: arn:aws:ecs:us-west-2:712793492101:task/39f3fe80-e8c7-409d-a6f4-d1af677b0695 is RUNNING.
                            dc30-qa2-web - INFO: Deploying new version to instance(s).
                  dc30-qa2-global-worker - INFO: Environment update is starting.
               dc30-qa2-callbacks-worker - INFO: Environment health has transitioned from Severe to Info. Application update in progress on 1 instance. 0 out of 1 instance completed (running for 65 seconds).
               dc30-qa2-callbacks-worker - INFO: New application version was deployed to running EC2 instances.
               dc30-qa2-callbacks-worker - INFO: Environment update completed successfully.
                  dc30-qa2-global-worker - INFO: Deploying new version to instance(s).
                            dc30-qa2-web - INFO: Starting new ECS task with awseb-dc30-qa2-web-n5vvhyeag6:33.
                            dc30-qa2-web - INFO: ECS task: arn:aws:ecs:us-west-2:712793492101:task/628e706c-fdda-4d9c-bed1-2a622eda90c2 is RUNNING.
                            dc30-qa2-web - INFO: New application version was deployed to running EC2 instances.
                            dc30-qa2-web - INFO: Environment update completed successfully.
                  dc30-qa2-global-worker - INFO: Starting new ECS task with awseb-dc30-qa2-global-worker-mw6wca2trw:25.
                  dc30-qa2-global-worker - INFO: ECS task: arn:aws:ecs:us-west-2:712793492101:task/19ee6116-a278-4674-a2dc-9d36a43630fb is RUNNING.
                  dc30-qa2-global-worker - INFO: New application version was deployed to running EC2 instances.
                  dc30-qa2-global-worker - INFO: Environment update completed successfully.
                  dc30-qa2-global-worker - INFO: Environment health has transitioned from Degraded to Info. Application update in progress. 1 out of 1 instance completed (running for 90 seconds).

```

### CODE REVIEW

File: lib/push_eb.py [hunk 1]
```diff
@@ -131,16 +183,24 @@ def deploy(self):
...
+      'deploy',
+      '--modules'
+    ] + self.config.env_list + [
+      '--env-group-suffix', hashlib.sha1(self.project_name()).hexdigest()[:10],
+      '--message', self.version,
+      '--region', self.region(),
+      '--profile', DEFAULT_PROFILE
+    ]
+    app = EBCli(argv=eb_args)
+    ebrun.run_app(app)
```
Comment (jeet4320) [inline lib/push_eb.py]: why do we need to call this

File: lib/run_script.py [hunk 2]
```diff
@@ -45,6 +40,16 @@ def run():
     finally:
       sys.exit(1)
 
+def run_script(script_name):
```
Comment (jeet4320) [inline lib/run_script.py]: why `run_script.py` need to be changed

Comment (whatisdot) [reply lib/run_script.py]: This isn't necessary for this PR and I can remove it. I added this in to skip exception handling, allowing me to see a stacktrace when there was a failure.

File: lib/push_eb.py (hunk 1 above)
Comment (whatisdot) [reply lib/push_eb.py]: *(For sake of transparency, I've uploaded the ebcli source to a private repository.)*

This is how the [ebcli initializes its commands](https://github.com/bodyshopbidsdotcom/ebcli_source/blob/master/core/ebcore.py#L164).

**Pull request merged**